pkg/config/release.go

```diff
@@ -163,6 +183,340 @@ func GetChangedTemplates(path, baseRev string) ([]ConfigMapSource, error) {
+// After all changes have been identified, the list gets "trimmed": all changed refs that are included in
+// changed refs are removed from the list since the refs will be tested when the chain gets tested. This
+// is done for changed chains included in changed workflows. This helps reduce the number of rehearsals
+// that are performed.
+func GetChangedRegistrySteps(path, baseRev string, refs map[string]api.LiteralTestStep, chains map[string][]api.TestStep, workflows map[string]api.MultiStageTestConfiguration) ([]RegistryChanges, error) {
+	revChanges, err := getRevChanges(path, RegistryPath, baseRev, true)
+	if err != nil {
+		return nil, err
+	}
+	var registryChanges []RegistryChanges
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load registry: %v", err)
+	}
+	fullRegistry := registry{refs: refs, chains: chains, workflows: workflows}
+	cToS := generateCommandToSteps(fullRegistry)
+	seenReference := make(map[string]bool)
+	for _, c := range revChanges {
+		if filepath.Ext(c.Filename) == ".yaml" {
+			change, err := loadRegistryChange(filepath.Join(path, RegistryPath), c.Filename)
+			if err != nil {
+				return nil, err
+			}
+			if change.Reference != nil {
+				if _, ok := seenReference[change.Reference.Reference.As]; ok {
+					continue
+				}
+				seenReference[change.Reference.Reference.As] = true
+			}
+			registryChanges = append(registryChanges, change)
+		} else if strings.HasSuffix(c.Filename, "commands.sh") {
+			stepFilename, err := getStepsFromCommands(c.Filename, c.SHA, cToS)
+			if err != nil {
+				logrus.Errorf("%v", err)
+			}
+			change, err := loadRegistryChange(filepath.Join(path, RegistryPath), stepFilename)
+			if err != nil {
+				return nil, err
+			}
+			if change.Reference != nil {
+				if _, ok := seenReference[change.Reference.Reference.As]; ok {
+					continue
+				}
+				seenReference[change.Reference.Reference.As] = true
+			}
+			registryChanges = append(registryChanges, change)
+		}
+	}
+	trimmedChanges := trimRegistryChanges(registryChanges, fullRegistry)
+	return trimmedChanges, nil
+}
+
+// unrollChains expands/unrolls all chains recursively into references.
+// Borrowed from pkg/resolver/resolver.go
+func unrollChains(input []api.TestStep, fullRegistry registry) (unrolledSteps []api.TestStep, errs []error) {
+	for _, step := range input {
+		if step.Chain != nil {
+			chain, ok := fullRegistry.chains[*step.Chain]
+			if !ok {
+				return []api.TestStep{}, []error{fmt.Errorf("unknown step chain: %s", *step.Chain)}
+			}
+			// handle nested chains
+			chain, err := unrollChains(chain, fullRegistry)
+			if err != nil {
+				errs = append(errs, err...)
+			}
+			unrolledSteps = append(unrolledSteps, chain...)
+			continue
+		}
+		unrolledSteps = append(unrolledSteps, step)
+	}
+	return
+}
+
+// getNestedChainsRecursive returns the names of all chains nested inside other chains in the provided RegistryChanges
+func getNestedChainsRecursive(chains []api.TestStep, fullRegistry registry) []string {
```

Comment:
nit: it's not necessary or useful to tell people this is recursive in the method name